src/core/border_router/routing_manager.cpp

```diff
@@ -365,85 +365,98 @@ void RoutingManager::EvaluateOmrPrefix(OmrPrefixArray &aNewOmrPrefixes)
 {
     NetworkData::Iterator           iterator = NetworkData::kIteratorInit;
     NetworkData::OnMeshPrefixConfig onMeshPrefixConfig;
-    Ip6::Prefix *                   electedOmrPrefix           = nullptr;
-    signed int                      electedOmrPrefixPreference = 0;
-    Ip6::Prefix *                   publishedLocalOmrPrefix    = nullptr;
+    OmrPrefix *                     favoredOmrEntry = nullptr;
+    OmrPrefix *                     localOmrEntry   = nullptr;
 
     OT_ASSERT(mIsRunning);
 
     while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, onMeshPrefixConfig) == kErrorNone)
     {
-        const Ip6::Prefix &prefix = onMeshPrefixConfig.GetPrefix();
+        OmrPrefix *entry;
 
         if (!IsValidOmrPrefix(onMeshPrefixConfig))
         {
             continue;
         }
 
-        if (aNewOmrPrefixes.Contains(prefix))
+        entry = aNewOmrPrefixes.FindMatching(onMeshPrefixConfig.GetPrefix());
+
+        if (entry != nullptr)
         {
-            // Ignore duplicate prefixes.
-            continue;
-        }
+            // Update the entry if we find the same prefix with higher
+            // preference in network data
 
-        if (aNewOmrPrefixes.PushBack(prefix) != kErrorNone)
+            if (onMeshPrefixConfig.GetPreference() <= entry->GetPreference())
+            {
+                continue;
+            }
+
+            entry->SetPreference(onMeshPrefixConfig.GetPreference());
+        }
+        else
         {
-            LogWarn("EvaluateOmrPrefix: Too many OMR prefixes, ignoring prefix %s", prefix.ToString().AsCString());
-            continue;
+            entry = aNewOmrPrefixes.PushBack();
+
+            if (entry == nullptr)
+            {
+                LogWarn("EvaluateOmrPrefix: Too many OMR prefixes, ignoring prefix %s",
+                        onMeshPrefixConfig.GetPrefix().ToString().AsCString());
+                break;
```

Comment:
```suggestion
                continue;
```
`continue` might be a better idea because we may still update the priorities of existing entries?

Reply:
Good suggestion. Changed in new push.